hamilton/experimental/h_schema.py

```diff
@@ -0,0 +1,277 @@
+        for key in d["added"]:
+            schema_metadata_diff[key] = "added"
+        for key in d["removed"]:
+            schema_metadata_diff[key] = "deleted"
+        for key in d["edited"]:
+            schema_metadata_diff[key] = diff_template.format(
+                ref=reference_schema.metadata[key], cur=current_schema.metadata[key]
+            )
+
+    for name in current_names.intersection(reference_names):
+        current_field = current_schema.field(name)
+        reference_field = reference_schema.field(name)
+        if current_field.equals(reference_field, check_metadata=check_field_metadata):
+            continue
+
+        field_diff = {}
+        if current_field.nullable != reference_field.nullable:
+            field_diff["nullable"] = diff_template.format(
+                ref=reference_field.nullable, cur=current_field.nullable
+            )
+        if not current_field.type.equals(reference_field.type):
+            field_diff["type"] = diff_template.format(
+                ref=reference_field.type, cur=current_field.type
+            )
+        if check_field_metadata:
+            field_metadata_diff = {}
+            d = dict_diff(current_field.metadata, reference_field.metadata)
+            for key in d["added"]:
+                field_metadata_diff[key] = "added"
+            for key in d["removed"]:
+                field_metadata_diff[key] = "deleted"
+            for key in d["edited"]:
+                field_metadata_diff[key] = diff_template.format(
+                    ref=reference_field.metadata[key], cur=current_field.metadata[key]
+                )
+
+        schema_diff[name] = field_diff
+
+    return schema_diff
+
+
+def dict_diff(current_map: Dict[str, str], reference_map: Dict[str, str]) -> dict:
+    """Generic diff of two mappings"""
+    current_only, reference_only, edit = [], [], []
+
+    for key, value1 in current_map.items():
+        value2 = reference_map.get(key)
+        if value2 is None:
+            current_only.append(key)
+            continue
+
+        if value1 != value2:
+            edit.append(key)
+
+    for key, value2 in reference_map.items():
+        value1 = current_map.get(key)
+        if value1 is None:
+            reference_only.append(key)
+            continue
+
+    return dict(added=current_only, removed=reference_only, edited=edit)
+
+
+@functools.singledispatch
+def _get_arrow_schema(df, allow_copy: bool = True) -> pyarrow.Schema:
+    # the property required for `from_dataframe()`
+    if not hasattr(df, "__dataframe__"):
+        raise NotImplementedError(f"Type {type(df)} is currently unsupported.")
+
+    # try to convert to Pyarrow using zero-copy; otherwise hit RuntimeError
+    try:
+        df = from_dataframe(df, allow_copy=False)
+    except RuntimeError as e:
+        if allow_copy is False:
+            raise e
+        df = from_dataframe(df, allow_copy=True)
+
+    return df.schema
+
+
+@_get_arrow_schema.register
```

Comment:
So yeah, this uses something similar to single dispatch (which we use elsewhere). We have a few too many ways of building polymorphic things, but again, need to internalize what this is getting (seems pretty cool here).